src/crystal/syntax_highlighter/html.cr

```diff
@@ -0,0 +1,98 @@
+require "../syntax_highlighter"
+
+# A syntax highlighter that renders Crystal source code with HTML markup.
+#
+# ```
+# code = %(foo = bar("baz\#{PI + 1}") # comment)
+# html = Crystal::SyntaxHighlighter::HTML.highlight(code)
+# html # => "foo <span class=\"o\">=</span> bar(<span class=\"s\">&quot;baz</span><span class=\"i\">\#{</span><span class=\"t\">PI</span> <span class=\"o\">+</span> <span class=\"n\">1</span><span class=\"i\">}</span><span class=\"s\">&quot;</span>) <span class=\"c\"># comment</span>"
+# ```
+class Crystal::SyntaxHighlighter::HTML < Crystal::SyntaxHighlighter
+  # Highlights *code* and writes the result to *io*.
+  def self.highlight(io : IO, code : String)
+    new(io).highlight(code)
+  end
+
+  # Highlights *code* and returns the result.
+  def self.highlight(code : String)
+    String.build do |io|
+      highlight(io, code)
+    end
+  end
+
+  # Highlights *code* or returns unhighlighted *code* on error.
+  #
+  # Same as `.highlight(code : String)` except that any error is rescued and
+  # returns unhighlighted source code.
+  def self.highlight!(code : String)
```

Comment:
What was the rationale to add an error-ignoring method? It wasn't there before and it seems kinda separate to the overall idea of the refactor.

Reply:
It's the other way around: Previously, `Crystal::Doc::Highlighter.highlight` would fall back to return the plain input if there's an error. The other method didn't exist before. It was added because a public API for this should have the ability to report errors.
I considered if maybe the logic of `highlight!` should not be part of the public API and only implemented in the compiler. But I figure it's not very specific to the compiler and it could be useful for other applications as well.